### PR TITLE
Enable Partition metric collection during ongoing execution to another k…

### PR DESCRIPTION
**Pull request by vmaheshw**
This PR resolves <https://github.com/linkedin/cruise-control/issues/1497>.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 1]
```diff
@@ -96,7 +127,18 @@ protected void fetchMetricsForAssignedPartitions() throws SamplingException {
...
       if (_sampleStore != null) {
-        _sampleStore.storeSamples(samples);
+        if (_samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
+          _sampleStore.storeSamples(new MetricSampler.Samples(new HashSet<>(), samples.brokerMetricSamples()));
+        } else {
+          _sampleStore.storeSamples(samples);
+        }
+      }
+
+      //replace _sampleStore
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:137]: Comment is not clear.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java [hunk 2]
```diff
@@ -123,14 +123,18 @@ private void updateCachedNumCoresByBroker(Cluster cluster) {
     // Get partition metric samples.
     Map<Integer, Integer> skippedPartitionByBroker = null;
     Set<PartitionMetricSample> partitionMetricSamples = new HashSet<>();
-    if (samplingMode == MetricSampler.SamplingMode.ALL || samplingMode == MetricSampler.SamplingMode.PARTITION_METRICS_ONLY) {
+    if (samplingMode == MetricSampler.SamplingMode.ALL ||
+        samplingMode == MetricSampler.SamplingMode.PARTITION_METRICS_ONLY ||
+        samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java:128]: Nit: Can we simplify this as:

```
samplingMode != MetricSampler.SamplingMode.BROKER_METRICS_ONLY
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java [hunk 3]
```diff
@@ -123,14 +123,18 @@ private void updateCachedNumCoresByBroker(Cluster cluster) {
...
       skippedPartitionByBroker = addPartitionMetricSamples(cluster, partitionsDotNotHandled, partitionMetricSamples);
     }
 
     // Get broker metric samples.
     int skippedBroker = 0;
     Set<BrokerMetricSample> brokerMetricSamples = new HashSet<>();
-    if (samplingMode == MetricSampler.SamplingMode.ALL || samplingMode == MetricSampler.SamplingMode.BROKER_METRICS_ONLY) {
+    if (samplingMode == MetricSampler.SamplingMode.ALL ||
+        samplingMode == MetricSampler.SamplingMode.BROKER_METRICS_ONLY ||
+        samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsProcessor.java:137]: Nit: Can we simplify this as:

```
samplingMode != MetricSampler.SamplingMode.PARTITION_METRICS_ONLY
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 4]
```diff
@@ -124,7 +166,9 @@ protected void fetchMetricsForAssignedPartitions() throws SamplingException {
     if (_samplingMode == MetricSampler.SamplingMode.ALL || _samplingMode == MetricSampler.SamplingMode.PARTITION_METRICS_ONLY) {
       usePartitionMetricSamples(samples.partitionMetricSamples());
     }
-    if (_samplingMode == MetricSampler.SamplingMode.ALL || _samplingMode == MetricSampler.SamplingMode.BROKER_METRICS_ONLY) {
+    if (_samplingMode == MetricSampler.SamplingMode.ALL ||
+        _samplingMode == MetricSampler.SamplingMode.BROKER_METRICS_ONLY ||
+        _samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:171]: Nit: Can we simplify this as:
```
_samplingMode != MetricSampler.SamplingMode.PARTITION_METRICS_ONLY
```

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 5]
```diff
@@ -96,7 +127,18 @@ protected void fetchMetricsForAssignedPartitions() throws SamplingException {
...
+        } else {
+          _sampleStore.storeSamples(samples);
+        }
+      }
+
+      //replace _sampleStore
+      if (_samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION && _sampleStoreForPartitionMetricsDuringExecution
+          != null) {
+        _sampleStoreForPartitionMetricsDuringExecution.storeSamples(new MetricSampler.Samples(samples.partitionMetricSamples(),
+            new HashSet<>()));
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:141]: Can we use `Collections.emptySet()` rather than the costly hashmap creation?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 6]
```diff
@@ -96,7 +127,18 @@ protected void fetchMetricsForAssignedPartitions() throws SamplingException {
     try {
       MetricSampler.Samples samples = fetchSamples();
       if (_sampleStore != null) {
-        _sampleStore.storeSamples(samples);
+        if (_samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
+          _sampleStore.storeSamples(new MetricSampler.Samples(new HashSet<>(), samples.brokerMetricSamples()));
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:131]: Can we use `Collections.emptySet()` rather than the costly hashmap creation?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java (hunk 5 above)
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:141]: Nit: Bad indentation

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 7]
```diff
@@ -57,9 +59,38 @@
                 Timer fetchTimer,
                 Meter fetchFailureRate,
                 MetricSampler.SamplingMode samplingMode) {
+    this(metricSampler, cluster, sampleStore, null, assignedPartitions, startTimeMs,
+        endTimeMs, metricDef, fetchTimer, fetchFailureRate, samplingMode);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:63]: Bad indentation -- please use the style file of the project.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 8]
```diff
@@ -57,9 +59,38 @@
...
+  }
+
+  /**
+   * @param metricSampler The sampler used to retrieve metrics.
+   * @param cluster The Kafka cluster.
+   * @param sampleStore Sample store to persist the fetched samples, or skip storing samples if {@code null}.
+   * @param sampleStoreForPartitionMetricsDuringExecution Sample store to persist the fetched partition samples during execution
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:70]: Can we specify the behavior when this is `{@code null}`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+ *    topic, default value is set to {@link #DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT}.</li>
+ *   <li>{@link #BROKER_SAMPLE_STORE_TOPIC_PARTITION_COUNT_CONFIG}: The config for the number of partition for Kafka broker sample store topic,
+ *   default value is set to {@link #DEFAULT_BROKER_SAMPLE_STORE_TOPIC_PARTITION_COUNT}.</li>
+ *   <li>{@link #MIN_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS_CONFIG}: The config for the minimal retention time for Kafka partition sample
+ *   store topic, default value is set to {@link #DEFAULT_MIN_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS}.</li>
+ *   <li>{@link #MIN_BROKER_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS_CONFIG}: The config for the minimal retention time for Kafka broker sample store
+ *   topic, default value is set to {@link #DEFAULT_MIN_BROKER_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS}.</li>
+ * </ul>
+ */
+public class KafkaWriteOnlySampleStore implements SampleStore {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java:58]: The name of this sample store is not fully accurate. The current name implies that it stores regular samples for later use similar to `ReadOnlyKafkaSampleStore`. But its intended use is to store partition metric samples during ongoing executions.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+  protected static final Duration PRODUCER_CLOSE_TIMEOUT = Duration.ofMinutes(3);
+  // Keep additional windows in case some of the windows do not have enough samples.
+  protected static final int ADDITIONAL_WINDOW_TO_RETAIN_FACTOR = 2;
+
+  protected static final short DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR = 2;
+  protected static final int DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT = 32;
+  protected static final int DEFAULT_BROKER_SAMPLE_STORE_TOPIC_PARTITION_COUNT = 32;
+  protected static final long DEFAULT_MIN_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS = TimeUnit.HOURS.toMillis(1);
+  protected static final long DEFAULT_MIN_BROKER_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS = TimeUnit.HOURS.toMillis(1);
+  protected static final String PRODUCER_CLIENT_ID = "KafkaCruiseControlSampleStoreProducer";
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java:69]: Why does the producer of this class has the same client id as the `KafkaSampleStore`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+  public static final String SAMPLE_STORE_TOPIC_REPLICATION_FACTOR_CONFIG = "sample.store.topic.replication.factor";
+  public static final String PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT_CONFIG = "partition.sample.store.topic.partition.count";
+  public static final String BROKER_SAMPLE_STORE_TOPIC_PARTITION_COUNT_CONFIG = "broker.sample.store.topic.partition.count";
+  public static final String MIN_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS_CONFIG = "min.partition.sample.store.topic.retention.time.ms";
+  public static final String MIN_BROKER_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS_CONFIG = "min.broker.sample.store.topic.retention.time.ms";
+
+  @Override
+  public void configure(Map<String, ?> config) {
+    _partitionMetricSampleStoreTopic = (String) config.get(PARTITION_METRIC_SAMPLE_STORE_TOPIC_CONFIG);
+    _brokerMetricSampleStoreTopic = (String) config.get(BROKER_METRIC_SAMPLE_STORE_TOPIC_CONFIG);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java:93]: I am confused. Why this this class use the same configs as `KafkaSampleStore`? Isn't this class supposed to generate samples that are collected during an ongoing execution to a new topic specific to partitions?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java (hunk 9 above)
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java:58]: I think this class needs some cleanup -- i.e. a lot of logic copied from `KafkaSampleStore` seems irrelevant and should be removed. If there is any remaining shared logic, we can move them to an Abstract class for them to share.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+          }
+        });
+      }
+    }
+    _producer.flush();
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Stored {} partition metric samples and {} broker metric samples to Kafka",
+          metricSampleCount.get(), brokerMetricSampleCount.get());
+    }
+  }
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java:254]: What is the intention here? Given that this is a draft PR, maybe the implementation of this class is in progress?

**Review by efeg — CHANGES_REQUESTED** [review]
Thanks for the PR! Left some comments.

Overall direction seems reasonable (i.e. having an additional sample store to keep partition metric samples during the ongoing execution). But there are issues around clarity and actual implementation logic -- unless I am missing something, the `KafkaWriteOnlySampleStore.java` does not seem to be doing what it is supposed to do (i.e. store samples given to it in a new topic). 

Given that this is a draft PR. I suspect the implementation of that class is in progress?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/MonitorConfig.java [hunk 13]
```diff
@@ -314,6 +315,15 @@ private MonitorConfig() {
       + "that persist the metric samples that have already been aggregated into Kafka Cruise Control. Later on the "
       + "persisted samples can be reloaded from the sample store to Kafka Cruise Control.";
 
+  /**
+   * <code>sample.partition.metrics.store.during.execution.class</code>
+   */
+  public static final String SAMPLE_PARTITION_METRICS_STORE_DURING_EXECUTION_CLASS_CONFIG = "sample.partition.metrics.store.during.execution.class";
+  public static final String DEFAULT_SAMPLE_PARTITION_METRICS_STORE_DURING_EXECUTION_CLASS = KafkaWriteOnlySampleStore.class.getName();
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/MonitorConfig.java]: I added similar to `DEFAULT_SAMPLE_STORE_CLASS`. I have fixed this one to `NoopSampleStore`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcherManager.java [hunk 14]
```diff
@@ -147,6 +147,23 @@ public boolean fetchMetricSamples(long startMs,
                                     long timeoutMs,
                                     SampleStore sampleStore,
                                     MetricSampler.SamplingMode samplingMode) {
+    return fetchMetricSamples(startMs, endMs, timeoutMs, sampleStore, null, samplingMode);
```
[the remark this answers is not in the document]
Comment (vmaheshw) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcherManager.java:151]: There is a null check before accessing this object.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java (hunk 9 above)
Comment (vmaheshw) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java]: Fixed.

Comment (vmaheshw) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java]: I was trying to create a generic store that can be used for both Broker metrics as well as Partition metrics. Fixed it to be specific to this use-case.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java (hunk 12 above)
Comment (vmaheshw) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaWriteOnlySampleStore.java]: Removed the broker metrics specific code.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleDuringExecutionStore.java [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+                       (recordMetadata, e) -> {
+                         if (e == null) {
+                           metricSampleCount.incrementAndGet();
+                         } else {
+                           LOG.error("Failed to produce partition metric sample for {} of timestamp {} due to exception",
+                               sample.entity().tp(), sample.sampleTime(), e);
+                         }
+                       });
+      }
+    }
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleDuringExecutionStore.java:192]: Duplicate code -- please see my comment on using abstract class to extract the common logic.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java [hunk 16]
```diff
@@ -96,7 +128,17 @@ protected void fetchMetricsForAssignedPartitions() throws SamplingException {
...
-        _sampleStore.storeSamples(samples);
+        if (_samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION) {
+          _sampleStore.storeSamples(new MetricSampler.Samples(Collections.emptySet(), samples.brokerMetricSamples()));
+        } else {
+          _sampleStore.storeSamples(samples);
+        }
+      }
+
+      if (_samplingMode == MetricSampler.SamplingMode.ONGOING_EXECUTION &&
+          _sampleStoreForPartitionMetricsDuringExecution != null) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/MetricFetcher.java:139]: Nit: The common convention is to carry `&&` to the new line. -- i.e. break before an operator.

This should be a CheckStyle rule -- I created to enforce that: https://github.com/linkedin/cruise-control/issues/1549

**Review by efeg — COMMENTED** [review]
Thanks for the updates!
Left some comments.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOngoingExecutionStore.java [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+    _partitionSampleStoreTopicPartitionCount = partitionSampleStoreTopicPartitionCountString == null
+                                               || partitionSampleStoreTopicPartitionCountString.isEmpty()
+                                               ? DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT
+                                               : Integer.parseInt(partitionSampleStoreTopicPartitionCountString);
+    String sampleStoreTopicRetentionTimeMsString = (String) config.get(
+        PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_RETENTION_TIME_MS_CONFIG);
+    _partitionSampleStoreTopicRetentionTimeMs = sampleStoreTopicRetentionTimeMsString == null
+                                                || sampleStoreTopicRetentionTimeMsString.isEmpty()
+                                                ? DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS
+                                                : Long.parseLong(sampleStoreTopicRetentionTimeMsString);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOngoingExecutionStore.java:70]: Looks like (1) `_partitionSampleStoreTopicPartitionCount` and (2) `_partitionSampleStoreTopicRetentionTimeMs` are used only once to ensure that the topic is created. So rather than keeping variables throughout the lifetime of the `KafkaPartitionMetricSampleOngoingExecutionStore` object, maybe we can just extract their values from config in `ensureTopicsCreated(Map<String, ?> config)` and pass them to `wrapTopic`.

-- Note: the same applies to `KafkaSampleStore`, ideally we should refactor it as well, but feel free to ignore this for that class as it was not introduced in this PR.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOngoingExecutionStore.java [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+ *   <li>{@link #PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_CONFIG}: The config for Kafka topic name to store partition samples.</li>
+ *   <li>{@link #PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_REPLICATION_FACTOR_CONFIG}: The config for the replication factor of Kafka
+ *   sample store topics, default value is set to {@link #DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR}.</li>
+ *   <li>{@link #PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_PARTITION_COUNT_CONFIG}: The config for the number of partition for Kafka
+ *   partition sample store topic, default value is set to {@link #DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT}.</li>
+ *   <li>{@link #PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_RETENTION_TIME_MS_CONFIG}: The config for the minimal retention time for
+ *   Kafka partition sample store topic, default value is set to {@link #DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS}.</li>
+ * </ul>
+ */
+public class KafkaPartitionMetricSampleOngoingExecutionStore extends AbstractKafkaSampleStore {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOngoingExecutionStore.java:33]: Nit: To match with the config names, should we rename this function to `KafkaPartitionMetricSampleOnExecutionStore`?

Note that with this change, "### KafkaPartitionMetricSampleOngoingExecutionStore configurations" title in the Configuration.md file should also be updated.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java [hunk 19]
```diff
@@ -36,7 +36,6 @@ public void configure(Map<String, ?> config) {
     for (int i = 0; i < numProcessingThreads; i++) {
       _consumers.add(createSampleStoreConsumer(config, CONSUMER_CLIENT_ID_PREFIX));
     }
-    _loadingProgress = -1.0;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java:39]: Wouldn't this change cause `_loadingProgress` to be unassigned? Should we instead have `_loadingProgress = LOADING_PROGRESS;` here?

Note: we can also delete the function `evictSamplesBefore(long timestamp)`, but feel free to ignore this, because it is not introduced in this PR.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/MonitorConfig.java [hunk 20]
```diff
@@ -311,9 +312,18 @@ private MonitorConfig() {
...
 
+  /**
+   * <code>sample.partition.metrics.store.ongoing.execution.class</code>
+   */
+  public static final String SAMPLE_PARTITION_METRIC_STORE_ONGOING_EXECUTION_CLASS_CONFIG = "sample.partition.metric.store.ongoing.execution.class";
+  public static final String DEFAULT_SAMPLE_PARTITION_METRIC_STORE_ONGOING_EXECUTION_CLASS = NoopSampleStore.class.getName();
+  public static final String SAMPLE_PARTITION_METRIC_STORE_ONGOING_EXECUTION_CLASS_DOC = "The sample store during execution class name. "
+      + "User may configure a sample store that persists the partition metric samples collected while there is an ongoing execution. "
+      + "This config is not intended for enhancing the fault tolerance of the system through recovery of the historical load information.";
+
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/config/constants/MonitorConfig.java:326]: Since some of the config names have changed from `ongoing` to `on`, should we use consistent name replacement from `ongoing` to `on` in all config names?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java [hunk 21]
```diff
@@ -9,7 +9,7 @@
 import java.util.Map;
 import java.util.concurrent.Executors;
 
-import static com.linkedin.kafka.cruisecontrol.monitor.sampling.SamplingUtils.createSampleStoreConsumer;
+import static com.linkedin.kafka.cruisecontrol.monitor.sampling.SamplingUtils.*;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/ReadOnlyKafkaSampleStore.java:12]: Nit: Can we avoid wildcard import?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOnExecutionStore.java [hunk 22]
```diff
@@ -58,30 +58,30 @@ public void configure(Map<String, ?> config) {
...
                                                || partitionSampleStoreTopicPartitionCountString.isEmpty()
                                                ? DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_PARTITION_COUNT
                                                : Integer.parseInt(partitionSampleStoreTopicPartitionCountString);
     String sampleStoreTopicRetentionTimeMsString = (String) config.get(
         PARTITION_METRIC_SAMPLE_STORE_ON_EXECUTION_TOPIC_RETENTION_TIME_MS_CONFIG);
-    _partitionSampleStoreTopicRetentionTimeMs = sampleStoreTopicRetentionTimeMsString == null
+    long partitionSampleStoreTopicRetentionTimeMs = sampleStoreTopicRetentionTimeMsString == null
                                                 || sampleStoreTopicRetentionTimeMsString.isEmpty()
                                                 ? DEFAULT_PARTITION_SAMPLE_STORE_TOPIC_RETENTION_TIME_MS
                                                 : Long.parseLong(sampleStoreTopicRetentionTimeMsString);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/KafkaPartitionMetricSampleOnExecutionStore.java]: Nit: indentation broken after name change.

**Review by efeg — COMMENTED** [review]
Thanks for all the updates!
Left 2 more nits otherwise LGTM.
